File: clone_list.c
#include <stdio.h>
#include <stdlib.h>
struct s_node {
        int           data;
        struct s_node *next;
        struct s_node *other;
    };

typedef struct s_node node;

node *init(int val)
{
	node *n = malloc(sizeof(node));
	n->data = val;
	n->next = NULL;
	n->other = NULL;
	return (n);
}

node *add(node *head, int val)
{
	node *el = init(val);
	while (head->next)
	{
		head = head->next;
	}
	head->next = el;
	return (el);
}

void printer(node *root)
{

	while (root)
	{
		printf("next: %d  ", root->data);
		if (!root->other)
			printf("other: NULL\n");
		else
			printf("other: %d\n", root->other->data);
		root = root->next;
	}
}

node *clone(node *head)
{
	node *cloned = init(head->data);
	node *copyHead = cloned;
	node *l2 = head->next;
	while (l2)
	{
		add(cloned, l2->data);
		l2 = l2->next;
	}
	return (copyHead);
}

int index_(node *root, node *el)
{
	int i = 0;
	if (!el)
		return(0);
	while (root)
	{
		if (root == el)
			return (i);
		i++;
		root = root->next;
	}
	return (-1);
}

node *pointer(node *root, int idx)
{
	if (idx == -1)
		return(NULL);
	while (root && idx > 0)
	{
		root = root->next;
		idx--;
	}
	return (root);
}

void connect(node *head, node *el, int i)
{
	el->other = pointer(head,i);
}



node *clone_list(node *root)
{
	node *l1 = root;
	node *l2 = clone(root);
	node *l2root = l2;
	node *l1root = root;
	while (l1 && l2)
	{
		node *tmp = l1->other;
		int idx = index_(l1root, tmp);
		connect(l2root, l2, idx);

		l1 = l1->next;
		l2 = l2->next;
	}
	return(l2root);
}



int main ()
{
	node *root = init(0);
	node *n1 = add(root,1);
	node *n2 = add(root,2);
	node *n3 = add(root,3);
	node *n4 = add(root,4);
	connect(root, root, 1);
    connect(root, n1, 2);
    connect(root, n2, 0);
    connect(root, n3, 1);
    connect(root, n4, 3);
	printer(root);
	printf("**************************************\n");
	
	node *new = clone_list(root);
	printf("**************************************\n");
	printer(new);

	return(0);

}

File: in_python/clonelist.py

class Node(object):
	def __init__(self,val):
		self.value = val
		self.next = None
		self.other = None

	def add(self,val):
		curr = self
		while(curr.next):
			curr = curr.next
		curr.next = Node(val)
		return curr.next

	def printer(self):
		while (self):
			print ("next: "+str(self.value))
			if (self.other):
				print("other:"+ str(self.other.value))
			else:
				print("other: NULL");
			self=self.next


	def clone(self):
		cloned = Node(self.value)
		headB = cloned
		headA= self.next
		while(headA):
			cloned.add(headA.value)
			headA = headA.next
		return(headB)

	def idx(self,elem):
		i = 0
		if (elem == None):
			return 0
		while (self):
			if (self == elem):
				return i
			i = i+1
			self = self.next
		return -1

	def ref(self, i):
		if (i == -1):
			return (None)
		while(self and i > 0):
			self = self.next
			i = i-1
		return(self)

	def direct(self, elem, i):
		 
		x = self.ref(i)
		elem.other = x

	def clone_list(self):
		l1 = self
		res = self.clone()
		l1root = l1
		resroot = res
		while (l1 and res):
			tmp = l1.other
			if not (tmp == None):
				i = l1root.idx(tmp)
				l1root.direct(res,i)

			l1 = l1.next
			res = res.next
		return(resroot)



lst = Node(1)
n1 = lst.add(2)
n2 = lst.add(3)
n3 = lst.add(4)
lst.direct(n1, 3)
lst.direct(n2, 0)

lst.printer()
print "++++++++++++"
lst2 = lst.clone_list()
lst2.printer()


